File: Web/10.SpringAPI/backend/src/main/java/ru/itmo/wp7/controller/UserController.java
package ru.itmo.wp7.controller;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.*;
import ru.itmo.wp7.domain.User;
import ru.itmo.wp7.exception.LoginInUseException;
import ru.itmo.wp7.service.UserService;
import ru.itmo.wp7.form.UserCredentials;

import javax.validation.Valid;
import java.util.List;
import java.util.Optional;

@RestController
@RequestMapping("/api/1")
public class UserController extends ApiController {
    @SuppressWarnings("SpringJavaAutowiredFieldsWarningInspection")
    @Autowired
    private UserService userService;

    @GetMapping("users/authenticated")
    public User getAuthenticatedUser(User user) {
        return user;
    }

    @GetMapping("user/{userId}")
    public Optional<User> getNotice(@PathVariable("userId") String userId) {
        return  userService.findById(Long.parseLong(userId));
    }

    @GetMapping("users")
    public List<User> getUsers() {
        return userService.findAll();
    }

    @PostMapping(path = "users")
    public void register(@Valid @RequestBody UserCredentials userCredentials)
            throws LoginInUseException {

        if (userService.countByLogin(userCredentials.getLogin()) != 0) {
            throw new LoginInUseException();
        }
        userService.register(userCredentials);
    }
}

File: software-design/reactive/src/main/kotlin/com/github/nothingelsematters/design/rx/controller/UserController.kt
package com.github.nothingelsematters.design.rx.controller

import com.github.nothingelsematters.design.rx.Product
import com.github.nothingelsematters.design.rx.User
import com.github.nothingelsematters.design.rx.service.UserService
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono

@RestController
@RequestMapping("/api/user")
class UserController(private val userService: UserService) {

    @PostMapping
    fun add(@RequestBody user: User): Mono<User> = userService.add(user)

    @GetMapping("{id}")
    fun get(@PathVariable id: Long): Mono<User?> = userService[id]

    @DeleteMapping("{id}")
    fun delete(@PathVariable id: Long) = userService.delete(id)

    @GetMapping("{id}/products")
    fun getProducts(@PathVariable id: Long): Flux<Product> = userService.getProducts(id)
}
